### PR TITLE
Improved error messages for when swiftc, clang, or sysroot can't be found

### PR DESCRIPTION
**Pull request by abertelrud**
No fix-its yet, but they will be added in further diffs.
https://bugs.swift.org/browse/SR-2271

### CODE REVIEW

File: Sources/Commands/Error.swift [hunk 1]
```diff
@@ -19,7 +19,7 @@ import var Utility.stderr
 
 public enum Error: Swift.Error {
     case noManifestFound
-    case invalidToolchain
+    case invalidToolchain(problem: String)
```
Comment (abertelrud) [inline Sources/Commands/Error.swift]: I realize that having a named property here is inconsistent with the other enum values; I will add names to them in a separate diff but wanted to keep this one small.  The reason for adding names will become more clear when we add a solution property as well.

**Comment by abertelrud** [comment]
I am going to do local checking on Linux and also check various error conditions before doing a smoke test and a merge.

**Comment by abertelrud** [comment]
I would like to add unit tests for this, but we don't seem to have any good structure for testing various invocations of the `swift` subcommands other than the particular case of building fixtures.  This is something to add.

**Comment by abertelrud** [comment]
@swift-ci please test

**Comment by abertelrud** [comment]
Linux failure is: fatal: unable to access 'https://github.com/apple/swift-corelibs-libdispatch.git/': Could not resolve host: github.com

File: Sources/Build/Command.compile(SwiftModule).swift [hunk 2]
```diff
@@ -14,7 +14,7 @@ import PackageLoading
 import Utility
 
 extension Command {
-    static func compile(swiftModule module: SwiftModule, configuration conf: Configuration, prefix: AbsolutePath, otherArgs: [String], SWIFT_EXEC: String) throws -> Command {
+    static func compile(swiftModule module: SwiftModule, configuration conf: Configuration, prefix: AbsolutePath, otherArgs: [String], compilerExec: AbsolutePath) throws -> Command {
```
Comment (ddunbar) [inline Sources/Build/Command.compile(SwiftModule).swift]: It seems like we should clean up this area to just pass the toolchain around...

**Comment by ddunbar** [comment]
LGTM, it is unfortunate we won't have test coverage of it though.

**Comment by ddunbar** [comment]
@swift-ci please test osx

File: Sources/Build/Command.compile(SwiftModule).swift (hunk 2 above)
Comment (abertelrud) [inline Sources/Build/Command.compile(SwiftModule).swift]: I agree.  I wanted to keep this diff minimal while still doing the basic work of cleaning up the error message, but it make sense to do it as a separate diff in the same PR.  I'll do that.

**Comment by abertelrud** [comment]
I agree.  I forgot where, but it one of the commit messages (I think) I mentioned the reason.  We need to add an infrastructure for testing environment settings and other external conditions (e.g. what `which` returns), as well as command line options.

We have a similar issue in not having testing of the command lines.  We definitely need to add testing of that but I didn't want to block this change on that.  I'll make sure that there is a JIRA, though.

**Comment by abertelrud** [comment]
One way would be to never call getenv() or shell out to `which` etc directly, but to always go through a protocol, which could then be mocked up in unit tests to return various boilerplate things.  Or we can invoke the built swift-build binary with a `PATH` that causes it to find a custom `which` that returns a mock value.  But it definitely requires some scaffolding.

**Comment by abertelrud** [comment]
I have updated to top-of-tree and will land after it passes tests (tests seem to get cancelled when you push an update).

**Comment by abertelrud** [comment]
@swift-ci please smoke test

**Comment by abertelrud** [comment]
@swift-ci please smoke test

**Pull request merged**